Exploration of preact bindings, logger and functions as actions

KidkArolis opened the pull request:
This is just a draft to see if there's any interest in any of the following features:

## Default logger
`{ log: true}` option that turns console logging for all actions, looks like this

![image](https://user-images.githubusercontent.com/324440/32344336-70b3ff52-bffe-11e7-9eb8-efa9148290ef.png)

## Preact bindings

`{ Provider, connect }` for Preact, means you can do this:

```js
const App = (atom, split) => (
  <div onClick={() => split('inc')}>
    {atom.count}
  </div>
)
export default connect()(App)
```

```js
  <Provider atom={atom}>
    <App />
  </Provider>
```

## Functions as actions

Functions as actions, this would mean you can organise your actions in modules and import them where needed. This one's a bit more controversial.

```js
const inc = () => split((get, split) => split({ count: get().count + 1})
export default connect()((atom, split) => <div onClick={inc}>atom.count</div>)
```

For me here are the pros/cons:

Pros

* you can keep your actions local to the component

Cons

* weird ceremonial syntax :-/ there are like 3 splits in that line
* 3 ways of doing stuff now, dispatching updates, dispatching actions and dispatching fns
* you lose the "action stream" concept, like tiny-atom follows this idea that there's a stream of actions which are plain objects and evolve consumes them and does stuff, here we break that...

KidkArolis committed:
Exploration of preact bindings, logger and functions as actions

KidkArolis commented:
On that last one, for me it's appealing the idea of keeping actions close to the component and plugging them in somehow. But not sure the proposed way here is the best. The only other way is to sort of "register" functions by importing them in your central `actions.js` file like so

```js
module.exports = Object.assign(
  {},
  require('./my-subapp/actions'),
  require('report/actions')
)
```

But that's fairly weird and you might have clashes. So then you like create constants file, so then you basically have redux, hm. So the nice thing about using functions as actions that it like doesn't matter, but then you lose this event stream idea basically unless you pretend the fn name is the action name.... but then you don't see the payloads, so yeah, not nice. Hm.

KwanMan commented on index.js:
```diff
@@ -28,9 +30,18 @@ module.exports = function createAtom (initialState, evolve, render, merge) {
     state = merge(state, nextState)
+    action = action || { payload: nextState }
+    if (log) {
+      console.log(`(${seq}) ${(action.fn ? action.fn.name : action.type) || '--'} { ${Object.keys(nextState || {}).join(', ')} }`, {
```
no template literals for es5

KwanMan commented on preact.js:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  Provider.prototype.render = function render (props) {
+    return props.children && props.children[0]
```
in preact world, `props.children` is always an array, an empty one if there is no children

KwanMan commented:
I've never really been a fan of function parameter overloading. I would've prefered explicit functions for mutating state and dispatching actions, so not a fan of adding a third way of using split

KwanMan commented:
```js
const inc = () => split((get, split) => split({ count: get().count + 1})
export default connect()((atom, split) => <div onClick={inc}>atom.count</div>)
```

I'm not really following this, where is that first split coming from?

KidkArolis commented:
Yeah, should have been:

```js
const inc = (get, split) => split({ count: get().count + 1})
export default connect()((atom, split) => <div onClick={() => split(inc)}>atom.count</div>)
```

But basically, not sure this is a good idea.. I think if you wanna go that route - you should just use `redux-zero`.

I think tiny-atom should remain closer to redux in the sense that it operates on a stream of actions and reduces them via it's singular evolve function. So no splitting of functions, only splitting of patches/actions is possible. I think that's fine.

The preact bindings are really nice idea, I've seen your preact-tiny-atom, but I'm basically thinking about just rolling it into this lib to make it more useful out of the box. Thinking of this now:

```js
const mapStateToProps = state => ({
  count: state.count,
  doubleCount: state.count * 2,
})

const mapSplitToProps = split => ({
  increment: x => split('increment', x),
  decrement: x => split('decrement', x)
})

const App = ({ count, doubleCount, increment, decrement }) => (
  <div>
    <div>{count}</div>
    <div>{doubleCount}</div>
    <button onClick={increment}>+</button>
    <button onClick={decrement}>-</button>
  </div>
)

export default connect(mapStateToProps, mapSplitToProps)(App)
```

And also thinking about introducing `options.debug` that gets called on both each action being split and on each state update, it should make a) debuggers/loggers easier to build (see #8) and b) make it clearer that it's just for debugging and c) make render signature cleaner.

Also - because tiny-atom is using the concept of actions just like redux, redux dev tools just work via `remotedev` integration, including time travel:

![image](https://user-images.githubusercontent.com/324440/32355440-98eb55ec-c026-11e7-982a-0d5b01f06d59.png)

I want to include that out of the box as well via `tiny-atom/devtool` or smth.

KidkArolis commented:
Like, redux is just .. bad api, it's just too confusing for everyone.. and not great defaults and DX..

Redux-zero is very cool, it just loses this mental model of actions. The benefit is that you bring actions closer to your components, which might be nice.

And tiny-atom, is closer to redux in that action stream sense, but it means actions are closer to the global store, actions are global. But I think that's ok. What you can do is import actions from your individual modules (just like combining redux reducers from multiple locations) or you can also create custom connector that registers actions for the component into tiny atom via context, it's doable as an extra, if that's your thing, while keeping the rest of tiny atom the same. Just need to write some docs for that..

KidkArolis committed:
Remove dev

KidkArolis committed:
Exploring further, redone loggging/debugging, mapActions, devtools

KidkArolis committed:
Nicen all things

KidkArolis commented:
Will close this to keep for the record. Will reopen a `next` branch.
